File: app/android/app/src/test/java/com/gymcoach/gym_coach/domain/PlanParseTest.kt
package com.gymcoach.gym_coach.domain

import com.gymcoach.gym_coach.domain.model.Plan
import com.google.common.truth.Truth.assertThat
import kotlinx.serialization.json.Json
import org.junit.Test

class PlanParseTest {

    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = false
    }

    @Test
    fun `parse end-to-end with training days and daily meals and start date`() {
        val raw = """
            {
              "weeks": 2,
              "weekly_structure": "x",
              "goal_summary": "y",
              "start_date": "2026-08-15T00:00:00Z",
              "training_days": [
                {
                  "week": 1,
                  "day_of_week": 1,
                  "title": "Upper",
                  "focus": "Push",
                  "estimated_minutes": 45,
                  "exercises": [
                    {"name": "Bench", "sets": 3, "reps": "8-12", "rest_seconds": 90}
                  ]
                }
              ],
              "daily_meals": [
                {
                  "week": 1,
                  "day_of_week": 1,
                  "total_kcal": 2000,
                  "protein_g": 150,
                  "carbs_g": 200,
                  "fat_g": 60,
                  "meals": [
                    {"slot": "breakfast", "name": "Oats", "kcal": 400, "protein_g": 25, "carbs_g": 60, "fat_g": 8, "ingredients": ["oats"]}
                  ]
                }
              ]
            }
        """.trimIndent()
        val plan = json.decodeFromString(Plan.serializer(), raw)
        assertThat(plan.weeks).isEqualTo(2)
        assertThat(plan.trainingDays).hasSize(1)
        assertThat(plan.trainingDays[0].exercises[0].name).isEqualTo("Bench")
        assertThat(plan.dailyMeals[0].meals[0].slot).isEqualTo("breakfast")
        assertThat(plan.startDate).isEqualTo("2026-08-15T00:00:00Z")
    }
}

File: app/test/domain/models/plan_parse_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/domain/models/plan.dart';

void main() {
  test('Plan parses valid LLM JSON output', () {
    final json = {
      'weeks': 12,
      'weekly_structure': '4 天训练 + 3 天休息',
      'goal_summary': '12 周增肌',
      'training_days': [
        {
          'week': 1,
          'day_of_week': 1,
          'title': '上肢推',
          'focus': '增肌',
          'estimated_minutes': 45,
          'exercises': [
            {'name': '哑铃卧推', 'sets': 4, 'reps': '8-12', 'rest_seconds': 90},
          ],
        },
      ],
      'daily_meals': [
        {
          'week': 1,
          'day_of_week': 1,
          'total_kcal': 2200,
          'protein_g': 160,
          'carbs_g': 220,
          'fat_g': 70,
          'meals': [
            {
              'slot': 'breakfast',
              'name': '燕麦',
              'kcal': 480,
              'protein_g': 30,
              'carbs_g': 60,
              'fat_g': 12,
              'ingredients': ['燕麦 50g'],
            },
          ],
        },
      ],
      'startDate': '2026-08-14T00:00:00.000Z',
    };
    final plan = Plan.fromJson(json);
    expect(plan.weeks, 12);
    expect(plan.startDate, isA<DateTime>());
    expect(plan.trainingDays.first.exercises.first.name, '哑铃卧推');
    expect(plan.dailyMeals.first.meals.first.slot, 'breakfast');
  });
}
